Physics System populates link bounding boxes

gabrielfpacheco opened the pull request:
# 🎉 New feature

Follow-up of #2787 

## Summary

Computes the link bounding boxes for all link entities that have `components::AxisAlignedBox` present. The computation follows the same logic as the models bbox, but falls back to the SDF calculation introduced in #2787 in case the feature is not present for the selected physics plugin.

## Test it

Run the following integration test:

```console
./build/gz-sim9/bin/INTEGRATION_physics_system --gtest_filter=*LinkBoundingBox*
```
It will load `bounding_boxes.sdf` world (see below) and perform link AABB checks while the models free-fall in the reach the ground.

![image](https://github.com/user-attachments/assets/01751ae7-e1dd-4ddb-a5f4-7a32f3e1c3f1)

> Note: after the previous PR is merged, this PR will rebase to `gz-sim9` and marked as ready to review

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

iche033 commented on test/integration/physics_system.cc:
```diff
@@ -1562,6 +1563,140 @@ TEST_F(PhysicsSystemFixture, GZ_UTILS_TEST_DISABLED_ON_WIN32(GetBoundingBox))
+  EXPECT_GT(duckBbox.Max().Y(), 0);
+}
+
+
```
remove empty line

gabrielfpacheco replied:
Amended https://github.com/gazebosim/gz-sim/pull/2821/commits/0595ea9c524fab6f75017b0c1954b52bdc970792 to address this

iche033 commented on src/Util.cc:
```diff
@@ -989,6 +989,38 @@ const common::Mesh *optimizeMesh(const sdf::Mesh &_meshSdf,
   return optimizedMesh;
 }
 
+math::AxisAlignedBox meshAxisAlignedBox(sdf::Mesh _sdfMesh)
```
consider also returning `std::optional` to indicate that the axis aligned box can not be computed

gabrielfpacheco replied:
Addressed in https://github.com/gazebosim/gz-sim/pull/2821/commits/84e70e798d227bac99f250b417344fec739833f2

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -4371,6 +4348,106 @@ void PhysicsPrivate::DisableContactSurfaceCustomization(const Entity &_world)
+          gzdbg << "Attempting to get a bounding box, but the physics "
+                 << "engine doesn't support feature "
+                 << "[GetLinkBoundingBox]. Link bounding boxes will be "
+                 << "computed from their collision shapes." << std::endl;
```
```suggestion
                 << "computed from their collision shapes based on their "
                 << "geometry properties in SDF." << std::endl;
```

gabrielfpacheco replied:
Amended https://github.com/gazebosim/gz-sim/pull/2821/commits/0595ea9c524fab6f75017b0c1954b52bdc970792 to address this.

iche033 reviewed (commented):
did a first pass review, overall apprach looks good

gabrielfpacheco commented on src/Util.cc:
```diff
@@ -989,6 +989,38 @@ const common::Mesh *optimizeMesh(const sdf::Mesh &_meshSdf,
   return optimizedMesh;
 }
 
+math::AxisAlignedBox meshAxisAlignedBox(sdf::Mesh _sdfMesh)
```
Oops.. Addressed in https://github.com/gazebosim/gz-sim/pull/2821/commits/84e70e798d227bac99f250b417344fec739833f2

gabrielfpacheco commented on include/gz/sim/Util.hh:
```diff
@@ -327,6 +328,20 @@ namespace gz
+    /// \param _sdfMesh Mesh SDF DOM.
+    /// \return The AABB of the mesh in its local frame.
+    GZ_SIM_VISIBLE math::AxisAlignedBox meshAxisAlignedBox(
+      const sdf::Mesh _sdfMesh);
```
Addressed in https://github.com/gazebosim/gz-sim/commit/84e70e798d227bac99f250b417344fec739833f2

gabrielfpacheco commented:
@iche033, I believe I addressed all your comments. If you agree with the proposed changes, could you please approve #2787 before? I would like to rebase this PR after the previous one is merged so that it focuses only on the physics system changes.

iche033 commented:
> @iche033, I believe I addressed all your comments. If you agree with the proposed changes, could you please approve #2787 before? I would like to rebase this PR after the previous one is merged so that it focuses only on the physics system changes.

Currently there's a build error as it needs the new API from sdformat. So I'm waiting for https://github.com/gazebosim/sdformat/pull/1547 to be merged first. We also need a new sdformat release afterwards in order to run CI in #2787 and make sure everything is green. We can only approve and merge a PR when the required CI checks are green.

gabrielfpacheco commented:
> So I'm waiting for [gazebosim/sdformat#1547](https://github.com/gazebosim/sdformat/pull/1547) to be merged first. We also need a new sdformat release afterwards in order to run CI in https://github.com/gazebosim/gz-sim/pull/2787 and make sure everything is green.

Yes, sure! I initially thought there might be a way to trigger the CI with dependent packages at specific branches, but then realized that the merge/release process is required, as you said.

Besides the dependency that prevents the proper CI checks, though, there are still some unresolved threads here from your first pass review, are those solved?

Regarding the sdformat PR, the CI is already green, and I believe all comments have been addressed. I've pinged the reviewers in https://github.com/gazebosim/sdformat/pull/1547#issuecomment-2752122608 recently, so I think we should be close to a merge.

iche033 commented on src/systems/physics/Physics.cc:
```diff
@@ -4371,6 +4348,107 @@ void PhysicsPrivate::DisableContactSurfaceCustomization(const Entity &_world)
+          gzdbg << "Attempting to get a bounding box, but the physics "
+                 << "engine doesn't support feature "
+                 << "[GetLinkBoundingBox]. Link bounding boxes will be "
+                 << "computed from their collision shapes based on their"
```
add space after `their`
```suggestion
                 << "computed from their collision shapes based on their "
```

iche033 commented:
> Besides the dependency that prevents the proper CI checks, though, there are still some unresolved threads here from your first pass review, are those solved?

yep looks good to me. Feel free to mark them as resolved once you address the comments.

iche033 commented on src/systems/physics/Physics.hh:
```diff
@@ -51,6 +51,7 @@
 
 #include <gz/sim/config.hh>
 #include <gz/sim/System.hh>
+#include <gz/sim/Link.hh>
```
can you move this include to `Physics.cc` given that it's only used there? and minor style nit: put it in alphabetical order with double quotes.

gabrielfpacheco commented:
@iche033, friendly ping :)

iche033 reviewed (approved):
looks good!

iche033 commented:
@mergify backport gz-sim8
